MPNet: Masked and Permuted Pre-training for Language Understanding

StillKeepTry opened the pull request:
# Model addition
[MPNet](https://arxiv.org/abs/2004.09297)

## Model description
MPNet introduces a novel self-supervised objective named masked and permuted language modeling for language understanding. It inherits the advantages of both the masked language modeling  (MLM) and the permuted language modeling (PLM) to addresses the limitations of MLM/PLM, and further reduce the inconsistency between the pre-training and fine-tuning paradigms.

# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

 albert, bert, XLM: @LysandreJik
 GPT2: @LysandreJik, @patrickvonplaten
 tokenizers: @mfuntowicz
 Trainer: @sgugger
 Benchmarks: @patrickvonplaten
 Model Cards: @julien-c
 examples/distillation: @VictorSanh
 nlp datasets: [different repo](https://github.com/huggingface/nlp)
 rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)
 Text Generation: @patrickvonplaten, @TevenLeScao
 Blenderbot, Bart, Marian, Pegasus: @patrickvonplaten
 T5: @patrickvonplaten
 Rag: @patrickvonplaten, @lhoestq
 EncoderDecoder: @patrickvonplaten
 Longformer, Reformer: @patrickvonplaten
 TransfoXL, XLNet: @TevenLeScao, @patrickvonplaten
 examples/seq2seq: @patil-suraj
 examples/bert-loses-patience: @JetRunner
 tensorflow: @jplu
 examples/token-classification: @stefan-it
 documentation: @sgugger
 FSTM: @stas00
 -->

patrickvonplaten commented:
Hey @StillKeepTry - could you maybe link the paper corresponding to your model add a small PR description? :-) That would be very helpful

patrickvonplaten commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+}
+
+
+class MPNetTokenizer(BertTokenizer):
```
Colud we maybe try to not inherit from `BertTokenizer` here? It seems like this class overwrites quite some methods - in this case it would be awesome if you could just copy past the required `BertTokenizer` functionality here. @LysandreJik @sgugger - what is your opinion on that?

LysandreJik replied:
Agreed, I don't see the point of inheriting from `BertTokenizer` here either.

StillKeepTry replied:
@patrickvonplaten I have removed inheritance and updated [tokenization_mpnet.py](https://github.com/StillKeepTry/transformers/blob/master/src/transformers/models/mpnet/tokenization_mpnet.py) and [tokenization_mpnet_fast.py](https://github.com/StillKeepTry/transformers/blob/master/src/transformers/models/mpnet/tokenization_mpnet_fast.py).  Can you have a check again?

patrickvonplaten commented on src/transformers/models/mpnet/configuration_mpnet.py:
```diff
@@ -0,0 +1,107 @@
+logger = logging.get_logger(__name__)
+
+MPNET_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "mpnet-base": "",
```
do we have pre-trained weights?

patrickvonplaten commented:
Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file?

JetRunner commented:
> Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file - it would be important to test the model :-) 
> 
> 
> 
> Also it would be amazing if you could give some context of MPNet - is there a paper, blog post, analysis, results that go along with the model? And are there pretrained weights? 
> 
> 
> 
> Thanks a lot! 

https://arxiv.org/abs/2004.09297

StillKeepTry commented:
> > Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file - it would be important to test the model :-)
> > Also it would be amazing if you could give some context of MPNet - is there a paper, blog post, analysis, results that go along with the model? And are there pretrained weights?
> > Thanks a lot!
> 
> https://arxiv.org/abs/2004.09297

ok

LysandreJik commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+        cls_token (:obj:`string`, `optional`, defaults to "<s>"):
+            The classifier token which is used when doing sequence classification (classification of the whole sequence
+            instead of per-token classification). It is the first token of the sequence when built with special tokens.
+        unk_token (:obj:`string`, `optional`, defaults to "<unk>"):
```
```suggestion
        unk_token (:obj:`string`, `optional`, defaults to "[UNK]"):
```

LysandreJik commented on src/transformers/models/mpnet/configuration_mpnet.py:
```diff
@@ -0,0 +1,107 @@
+    Configuration objects inherit from :class:`~transformers.PretrainedConfig` and can be used to control the model
+    outputs. Read the documentation from :class:`~transformers.PretrainedConfig` for more information.
+
+    Args:
+        vocab_size: Vocabulary size of `inputs_ids` in `MPNetModel`.
+        hidden_size: Size of the encoder layers and the pooler layer.
+        num_hidden_layers: Number of hidden layers in the Transformer encoder.
+        num_attention_heads: Number of attention heads for each attention layer in
+            the Transformer encoder.
+        intermediate_size: The size of the "intermediate" (i.e., feed-forward)
+            layer in the Transformer encoder.
+        hidden_act: The non-linear activation function (function or string) in the
+            encoder and pooler. If string, "gelu", "relu", "swish" and "gelu_new" are supported.
+        hidden_dropout_prob: The dropout probabilitiy for all fully connected
+            layers in the embeddings, encoder, and pooler.
+        attention_probs_dropout_prob: The dropout ratio for the attention
+            probabilities.
+        max_position_embeddings: The maximum sequence length that this model might
+            ever be used with. Typically set this to something large just in case (e.g., 512 or 1024 or 2048).
+        initializer_range: The sttdev of the truncated_normal_initializer for
+            initializing all weight matrices.
+        layer_norm_eps: The epsilon used by LayerNorm.
+        relative_attention_num_buckets: The buckets for relative positional embedding.
```
It would be great if this could be formatted like our other configuration files. You can probably copy-paste a lot from BERT: https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/configuration_bert.py#L62-L100

LysandreJik commented on src/transformers/models/mpnet/modeling_mpnet.py:
```diff
@@ -0,0 +1,1055 @@
+import math
+import os
+import warnings
+import logging
```
No need for this, you're importing transformers' logging utility already.

LysandreJik commented on src/transformers/models/mpnet/modeling_mpnet.py:
```diff
@@ -0,0 +1,1055 @@
+        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
+
+    def forward(self, input_ids=None, position_ids=None, inputs_embeds=None, **kwargs):
+        # Copied from transformers.modeling_roberta.RobertaEmbeddings.forward
```
Should this be placed above the method? @sgugger

LysandreJik commented on src/transformers/models/mpnet/modeling_mpnet.py:
```diff
@@ -0,0 +1,1055 @@
+"""
+
+MPNET_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary. Indices can be obtained using
+            :class:`transformers.MPNetTokenizer`. See :func:`transformers.PreTrainedTokenizer.encode` and
+            :func:`transformers.PreTrainedTokenizer.encode_plus` for details. `What are input IDs?
+            <../glossary.html#input-ids>`__
+        attention_mask (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`, defaults to :obj:`None`):
+            Mask to avoid performing attention on padding token indices. Mask values selected in ``[0, 1]``: ``1`` for
+            tokens that are NOT MASKED, ``0`` for MASKED tokens. `What are attention masks?
+            <../glossary.html#attention-mask>`__
+        position_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`, defaults to :obj:`None`):
+            Indices of positions of each input sequence tokens in the position embeddings. Selected in the range ``[0,
+            config.max_position_embeddings - 1]``. `What are position IDs? <../glossary.html#position-ids>`_
+        head_mask (:obj:`torch.FloatTensor` of shape :obj:`(num_heads,)` or :obj:`(num_layers, num_heads)`, `optional`, defaults to :obj:`None`):
+            Mask to nullify selected heads of the self-attention modules. Mask values selected in ``[0, 1]``: :obj:`1`
+            indicates the head is **not masked**, :obj:`0` indicates the head is **masked**.
+        inputs_embeds (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, hidden_size)`, `optional`, defaults to :obj:`None`):
+            Optionally, instead of passing :obj:`input_ids` you can choose to directly pass an embedded representation.
+            This is useful if you want more control over how to convert `input_ids` indices into associated vectors
+            than the model's internal embedding lookup matrix.
+        output_attentions (:obj:`bool`, `optional`):
+            Whether or not to return the attentions tensors of all attention layers. See ``attentions`` under returned
+            tensors for more detail.
+        output_hidden_states (:obj:`bool`, `optional`):
+            Whether or not to return the hidden states of all layers. See ``hidden_states`` under returned tensors for
+            more detail.
+        return_dict (:obj:`bool`, `optional`):
+            Whether or not to return a :class:`~transformers.file_utils.ModelOutput` instead of a plain tuple.
```
Could you update this to be more in line with the other models' documentation? You can copy-paste a lot from BERT's: https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_bert.py#L691-L738

LysandreJik commented on src/transformers/models/mpnet/modeling_mpnet.py:
```diff
@@ -0,0 +1,1055 @@
+        self.out_proj = nn.Linear(config.hidden_size, config.num_labels)
+
+    def forward(self, features, **kwargs):
+        x = features[:, 0, :]  # take <s> token (equiv. to [CLS])
```
```suggestion
        x = features[:, 0, :]  # take <s> token (equiv. to BERT's [CLS] token)
```

LysandreJik commented on src/transformers/models/mpnet/modeling_tf_mpnet.py:
```diff
@@ -0,0 +1,1305 @@
+""" TF 2.0 MPNet model. """
```
Could you add the copyright here?

LysandreJik reviewed (commented):
This is exceptionally clean, thank you! I've left a few comments regarding the documentation.

Also,  we've recently updated the way to process the inputs in TensorFlow. Maybe @jplu can give you some tips or help you implement the changes.

As @patrickvonplaten mentioned, the two test files `test_modeling_mpnet.py` and `test_modeling_tf_mpnet.py` are necessary for us to merge this. Let us know if you need help implementing those and we'll be sure to give you pointers.

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+    Args:
+        vocab_file (:obj:`str`):
+            Path to the vocabulary file.
+        bos_token (:obj:`string`, `optional`, defaults to "<s>"):
```
```suggestion
        bos_token (:obj:`string`, `optional`, defaults to :obj:`"<s>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+            Path to the vocabulary file.
+        bos_token (:obj:`string`, `optional`, defaults to "<s>"):
+            The beginning of sequence token that was used during pre-training. Can be used a sequence classifier token.
+            .. note:: When building a sequence using special tokens, this is not the token that is used for the
+            beginning of sequence. The token used is the :obj:`cls_token`.
```
```suggestion

            .. note::
            
                When building a sequence using special tokens, this is not the token that is used for the
                beginning of sequence. The token used is the :obj:`cls_token`.
```
This needs a bit more new lines to be properly formatted.

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+            The beginning of sequence token that was used during pre-training. Can be used a sequence classifier token.
+            .. note:: When building a sequence using special tokens, this is not the token that is used for the
+            beginning of sequence. The token used is the :obj:`cls_token`.
+        eos_token (:obj:`string`, `optional`, defaults to "</s>"):
```
```suggestion
        eos_token (:obj:`string`, `optional`, defaults to :obj:`"</s>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+            .. note:: When building a sequence using special tokens, this is not the token that is used for the
+            beginning of sequence. The token used is the :obj:`cls_token`.
+        eos_token (:obj:`string`, `optional`, defaults to "</s>"):
+            The end of sequence token. .. note:: When building a sequence using special tokens, this is not the token
+            that is used for the end of sequence. The token used is the :obj:`sep_token`.
```
```suggestion
            The end of sequence token. 
            
            .. note::
            
                When building a sequence using special tokens, this is not the token
                that is used for the end of sequence. The token used is the :obj:`sep_token`.
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+        eos_token (:obj:`string`, `optional`, defaults to "</s>"):
+            The end of sequence token. .. note:: When building a sequence using special tokens, this is not the token
+            that is used for the end of sequence. The token used is the :obj:`sep_token`.
+        sep_token (:obj:`string`, `optional`, defaults to "</s>"):
```
```suggestion
        sep_token (:obj:`string`, `optional`, defaults to :obj:`"</s>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+            The separator token, which is used when building a sequence from multiple sequences, e.g. two sequences for
+            sequence classification or for a text and a question for question answering. It is also used as the last
+            token of a sequence built with special tokens.
+        cls_token (:obj:`string`, `optional`, defaults to "<s>"):
```
```suggestion
        cls_token (:obj:`string`, `optional`, defaults to :obj:`"<s>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+        cls_token (:obj:`string`, `optional`, defaults to "<s>"):
+            The classifier token which is used when doing sequence classification (classification of the whole sequence
+            instead of per-token classification). It is the first token of the sequence when built with special tokens.
+        unk_token (:obj:`string`, `optional`, defaults to "[UNK]"):
```
```suggestion
        unk_token (:obj:`string`, `optional`, defaults to :obj:`"[UNK]"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+        unk_token (:obj:`string`, `optional`, defaults to "[UNK]"):
+            The unknown token. A token that is not in the vocabulary cannot be converted to an ID and is set to be this
+            token instead.
+        pad_token (:obj:`string`, `optional`, defaults to "<pad>"):
```
```suggestion
        pad_token (:obj:`string`, `optional`, defaults to :obj:`"<pad>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet.py:
```diff
@@ -0,0 +1,194 @@
+            token instead.
+        pad_token (:obj:`string`, `optional`, defaults to "<pad>"):
+            The token used for padding, for example when batching sequences of different lengths.
+        mask_token (:obj:`string`, `optional`, defaults to "<mask>"):
```
```suggestion
        mask_token (:obj:`string`, `optional`, defaults to :obj:`"<mask>"`):
```

sgugger commented on src/transformers/models/mpnet/tokenization_mpnet_fast.py:
```diff
@@ -0,0 +1,169 @@
+    Args:
+        vocab_file (:obj:`str`):
+            Path to the vocabulary file.
+        bos_token (:obj:`string`, `optional`, defaults to "<s>"):
```
Same comments as for the slow tokenizer (I think it will just need a copy-paste once you apply the suggestions in the slow tokenizer).

sgugger reviewed (approved):
This is very clean! Just have a few last nits on the docstrings of the tokenizers (and as Lysandre and Patrick mentioned, we need the test modeling files to merge this).
Thanks a lot for all your work!

sgugger commented:
Oh and another thing to do after the merge will be to add your new model to the main README and the documentation so that people can use it! The template should give you a file for the `.rst` (or you can use `docs/model_doc/bert.rst` as an example).

jplu reviewed (approved):
Thanks a lot!! This is a very good work! I'm totally ok with the TF part once there will be a test file added.

StillKeepTry commented:
I have updated `test_modeling_mpnet.py` now.

StillKeepTry commented:
Hi, every reviewer. I have fixed previous comments (like doc, format, and so on) and updated the `tokenization_mpnet.py` and `tokenization_mpnet_fast.py` by removing the inheritance. Besides, I also update `test_modeling_mpnet.py`, `test_modeling_tf_mpnet.py` for testing, and model weights into the model hub. 

So can you make some new reviews based on the current version? Thank you very much.

StillKeepTry commented:
Hi, every reviewer. Thank you for your valuable reviews. I have fixed previous comments (like doc, format, and so on) and updated the `tokenization_mpnet.py` and `tokenization_mpnet_fast.py` by removing the inheritance. Besides, I also upload test files (`test_modeling_mpnet.py`, `test_modeling_tf_mpnet.py`) for testing, and model weights into the model hub.

LysandreJik commented:
Fantastic, thanks for working on it! Will review today.

StillKeepTry commented:
@patrickvonplaten Hi, are there any new comments?

jplu commented:
Hello!

Still some comments:

1. Update the inputs handling in the TF file, we have merged an update for the booleans. You can see an example in the TF BERT file if you need one.
2. rebase and fix the conflicting files.
3. Fix the check_code_quality test.

patrickvonplaten reviewed (approved):
LGTM, if possible (if there are pre-trained weights) it would be great to also add an integration test with a pre-trained weight, similar to how it's done for roberta here: https://github.com/huggingface/transformers/blob/28c77ddf3bd39f3e528f41d94a67b88ad967173a/tests/test_modeling_roberta.py#L426.

patrickvonplaten commented:
I think something went wrong with the merge here :-/ Could you try to open a new PR that does not include all previous commits or fix this one?

StillKeepTry commented:
> I think something went wrong with the merge here :-/ Could you try to open a new PR that does not include all previous commits or fix this one?

OK, it seems something wrong when I update to the latest version. :(

StillKeepTry commented:
@patrickvonplaten @JetRunner @sgugger @LysandreJik @jplu The new PR is moved to [https://github.com/huggingface/transformers/pull/8971](https://github.com/huggingface/transformers/pull/8971)
